JoinColony/colonyCDapp
# Send notifications on simple payment action

## Description

Send notifications when a simple payment action is made via permissions.

Most of the work is done in this block ingestor PR: https://github.com/JoinColony/block-ingestor/pull/268, so please also review that.

Notifications should be sent to everybody in the colony who has notifications turned on and has not muted that colony. Another notification should be sent to the recipient of the funds.

## Testing

We need to check that the notifications are sent to the right people. Make sure that you start this testing with a clean environment.

**IMPORTANT:** You need the magicbell api key and secret to be in your .env file.

* Log into the app as Leela, and then run the following query in `localhost:20002` to check that notifications data is made for this user.

```graphql
query MyQuery {
  getUser(id: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461") {
    id
    profile {
      displayName
    }
    notificationsData {
      magicbellUserId
      userAddress
      mutedColonyAddresses
      notificationsDisabled
    }
  }
}
```

<img width="2558" alt="Screenshot 2024-09-03 at 17 15 32" src="https://github.com/user-attachments/assets/c6c29e10-33a4-45db-bf04-158d1acb3736">

* Create a simple payment action with Leela as the recipient.

<img width="2558" alt="Screenshot 2024-09-03 at 17 16 49" src="https://github.com/user-attachments/assets/37a60ed2-9011-4642-9132-1b6a80e91e91">

* Go to the Magicbell dashboard and check that two notifications were sent, each with Leela as the only recipient:

<img width="2558" alt="Screenshot 2024-09-03 at 18 25 33" src="https://github.com/user-attachments/assets/c72cf7e4-ad46-4266-8483-b2563b17c059">
<img width="2558" alt="Screenshot 2024-09-03 at 18 25 40" src="https://github.com/user-attachments/assets/84a7a813-8a09-49b5-89e8-040f5453c063">

* In an incognito window, log in as Amy (this will make a notifications user for her).

* Now make another simple payment action with Leela as the recipient, and again check the Magicbell dashboard. This time two Notifications should have been made, but one of them should have two recipients (Leela and Amy) while the `mention` notification should still only be going to Leela.

<img width="2558" alt="Screenshot 2024-09-03 at 18 29 13" src="https://github.com/user-attachments/assets/cb4ffc71-3c08-4219-89ce-64a9a8bc12b5">
<img width="2558" alt="Screenshot 2024-09-03 at 18 29 15" src="https://github.com/user-attachments/assets/7698b08c-6af0-4536-ad29-6aec3c5206bb">

* Run the following mutation to disable notifications for Leela:

```graphql
mutation MyMutation {
  updateNotificationsData(
    input: {userAddress: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461", notificationsDisabled: true}
  ) {
    userAddress
    notificationsDisabled
  }
}
```

* Make another simple payment with Leela as the recipient, and again check the Magicbell dashboard. This time only one notification should have been sent - the colony wide one - with Amy as the only recipient.

<img width="2558" alt="Screenshot 2024-09-03 at 18 32 58" src="https://github.com/user-attachments/assets/0c1b9ebf-e67f-4c1a-b603-4eb6227bc234">

* Run the following mutation to mute the planex colony for Amy:

```graphql
mutation MyMutation {
  updateNotificationsData(
    input: {userAddress: "0x9dF24e73f40b2a911Eb254A8825103723E13209C", mutedColonyAddresses: "0x59E0fe5e32A300Ebcb14787c4C198087559C36Ec"}
  ) {
    userAddress
    mutedColonyAddresses
  }
}
```

* Make a final simple payment with Leela as the recipient. This time NO notifications at all should be made.

## Diffs

**New stuff** ✨

* New utils for handling notifications in the block ingestor

**Changes** 🏗

* Changed the model for NotificationsData in amplify

Resolves #3010
